Added oracle content [#1881] (#1965)
ethereum/ethereum-org-website: src/content/developers/docs/oracles/index.md

> Oracles are a data feed that connect Ethereum to off-chain, real-world information, so you can query data in your smart contracts. For example, prediction market dapps use oracles to settle payments based on events. A prediction market may ask you to bet your ETH on the next president of the United States. They'll use an oracle to confirm the outcome and pay out to the winners.

Reviewer: ```suggestion
Oracles are data feeds that connect Ethereum to off-chain, real-world information, so you can query data in your smart contracts. For example, prediction market dapps use oracles to settle payments based on events. A prediction market may ask you to bet your ETH on the next president of the United States. They'll use an oracle to confirm the outcome and pay out to the winners.
```

## Prerequisites {#prerequisites}

Make sure you're familiar with [nodes](/developers/docs/nodes-and-clients/), [consensus mechanisms](/developers/docs/consensus-mechanisms/), and [smart contract anatomy](/developer/docs/smart-contracts/anatomy/), specifically events.

## What is an oracle {#what-is-an-oracle}

> Oracles act as a bridge between the blockchain and the real world. They allow you to query APIs to get information into your smart contracts. This could be anything from price information to weather reports.

Reviewer: ```suggestion
An oracle is a bridge between the blockchain and the real world. They act as on-chain APIs you can query to get information into your smart contracts. This could be anything from price information to weather reports.
```

## Why are they needed? {#why-are-they-needed}

> With a blockchain like Ethereum you need every node in the network to be able to replay every transaction and end up with the same result, guaranteed. APIs introduce potentially variable data. If you were sending someone an amount of ETH based on an agreed \$USD value using a price API, the query would return a different result from one day to the next. Not to mention, the API could be hacked or deprecated. If this happens, the nodes in the network wouldn't be able to agree on Ethereum's current state, effectively breaking [consensus](/developers/docs/consensus-mechanisms/).

Reviewer: ```suggestion
With a blockchain like Ethereum you need every node in the network to be able to replay every transaction and end up with the same result, guaranteed. APIs introduce potentially variable data. If you were sending someone an amount of ETH based on an agreed $USD value using a price API, the query would return a different result from one day to the next. Not to mention, the API could be hacked or deprecated. If this happens, the nodes in the network wouldn't be able to agree on Ethereum's current state, effectively breaking [consensus](/developers/docs/consensus-mechanisms/).
```

> With a blockchain like Ethereum you need every node in the network to be able to replay every transaction and end up with the same result, guaranteed. APIs introduce potentially variable data. If you were sending someone an amount of ETH based on an agreed \$USD value using a price API, the query would return a different result from one day to the next. Not to mention, the API could be hacked or deprecated. If this happens, the nodes in the network wouldn't be able to agree on Ethereum's current state, effectively breaking [consensus](/developers/docs/consensus-mechanisms/).

Reviewer: ```suggestion
With a blockchain like Ethereum you need every node in the network to be able to replay every transaction and end up with the same result, guaranteed. APIs introduce potentially variable data. If you were sending someone an amount of ETH based on an agreed $USD value using a price API, the query would return a different result from one day to the next. Not to mention, the API could be hacked or deprecated. If this happens, the nodes in the network wouldn't be able to agree on Ethereum's current state, effectively breaking [consensus](/developers/docs/consensus-mechanisms/).
```

> Oracles solve this problem by posting the data on the blockchain. So any node replaying the transaction will use the same immutable data that's posted for all to see. To do this, an oracle is made up of a smart contract and some off-chain components that can query APIs.

Reviewer: ```suggestion
Oracles solve this problem by posting the data on the blockchain. So any node replaying the transaction will use the same immutable data that's posted for all to see. To do this, an oracle is typically made up of a smart contract and some off-chain components that can query APIs, then periodically send transactions to update the smart contract's data.
```

> 

Reviewer: I think it'd be great (perhaps in a v2) to have a "use cases" section here. A few that come to mind:
- Price feeds (e.g. https://medium.com/compound-finance/announcing-compound-open-oracle-development-cff36f06aad3 https://chain.link/solutions/defi)
- Randomness generation (e.g. https://blog.chain.link/verifiable-random-functions-vrf-random-number-generation-rng-feature/ https://chain.link/solutions/chainlink-vrf)
- Proof of reserve (e.g. https://cointelegraph.com/news/1b-in-wrapped-bitcoin-now-being-audited-using-chainlink-s-proof-of-reserve)

## Oracle architecture {#oracle-architecture}

To understand how an oracle works, let's play through a scenario where your smart contract needs to know who won the superbowl. This is an example of how it could work:

1. Your smart contract requests information from an oracle smart contract.
2. The oracle smart contract emits an [event](/developers/docs/smart-contracts/anatomy/#events-and-logs).
3. Off-chain oracle nodes listen for events and upon hearing one, they query an API.
4. The API returns a JSON response to the nodes.
5. The nodes call on the oracle smart contract.

> 6. The oracle smart contract returns the data to your smart contract.

Reviewer: This is all within a single Ethereum tx? I don't think that's actually how it works. Let's discuss tomorrow 😄

### Security {#security}

> Notice in the above example there are multiple nodes. This is to prevent an oracle becoming a security risk. If a contract relies on one API or one off-chain node then there is a single point of failure. Because they're off-chain, they can be compromised. Often the oracle smart contract will require a set number of the same results from a whole network of nodes to improve security.

Reviewer: I would start the security section by explaining "An oracle is only as secure as its data source(s)...". 

An example might help, e.g. if an dapp uses Uniswap as an oracle for its ETH/DAI price feed, all an attacker has to do it move the price on Uniswap in order to manipulate/fool the dapp's understanding of the current price.
An example of how to combat this is a [feed system like MakerDAO's](https://developer.makerdao.com/feeds/), which which collates price data from a number of external price feeds instead of just relying on a single source.

> Notice in the above example there are multiple nodes.

As in multiple sources? I'm not clear on what "multiple nodes" refers to here - Ethereum nodes? Multiple Ethereum nodes doesn't ensure the security of oracles. Multiple _data sources_ can help improve security.

> Because they're off-chain, they can be compromised.

I think that statement is misleading. Even if it's on-chain, a feed can be compromised - it goes back to my point above about what the data source is & how many points of failure there are. An on-chain oracle might still have a single point of failure.

Author: The example uses off-chain oracle nodes to provided multiple sources for the data.
